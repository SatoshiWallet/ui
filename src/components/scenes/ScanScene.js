// @flow

import React, { Component, Fragment } from 'react'
import { ActivityIndicator, Text, TouchableHighlight, View } from 'react-native'
import OpenAppSettings from 'react-native-app-settings'
import { RNCamera } from 'react-native-camera'
// $FlowFixMe
import slowlog from 'react-native-slowlog'
import FAIcon from 'react-native-vector-icons/FontAwesome'
import Ionicon from 'react-native-vector-icons/Ionicons'

import SecondaryModal from '../../connectors/SecondaryModalConnector.js'
import s from '../../locales/strings.js'
import T from '../../modules/UI/components/FormattedText/index'
import { type PermissionStatus } from '../../reducers/PermissionsReducer.js'
import styles, { styles as styleRaw } from '../../styles/scenes/ScaneStyle'
import { scale } from '../../util/scaling.js'
import { SceneWrapper } from '../common/SceneWrapper.js'

type Props = {
  cameraPermission: PermissionStatus,
  torchEnabled: boolean,
  scanEnabled: boolean,
  deepLinkPending: boolean,
  deepLinkUri: string | null,
  qrCodeScanned: (data: string) => void,
  parseScannedUri: (data: string) => void,
  toggleEnableTorch: () => void,
  toggleAddressModal: () => void,
  toggleScanToWalletListModal: () => void,
  markAddressDeepLinkDone: () => any
}

const HEADER_TEXT = s.strings.send_scan_header_text

const DENIED_PERMISSION_TEXT = s.strings.scan_camera_permission_denied // blank string because way off-centered (not sure reason why)
const OPEN_SETTINGS_TEXT = s.strings.open_settings
// const TRANSFER_TEXT = s.strings.fragment_send_transfer
const ADDRESS_TEXT = s.strings.fragment_send_address
const FLASH_TEXT = s.strings.fragment_send_flash

export class Scan extends Component<Props> {
  constructor (props: Props) {
    super(props)

    this.checkForDeepLink()

    slowlog(this, /.*/, global.slowlogOptions)
  }

  componentDidUpdate () {
    this.checkForDeepLink()
  }

  checkForDeepLink () {
    if (this.props.deepLinkUri && this.props.deepLinkPending) {
      const deepLinkUri = this.props.deepLinkUri
      this.props.markAddressDeepLinkDone()
      this.props.parseScannedUri(deepLinkUri)
    }
  }

  render () {
    return (
      <Fragment>
        <SceneWrapper background="header" hasTabs={false}>
          {this.renderCameraArea()}
          <View style={styles.overlayButtonAreaWrap}>
            <TouchableHighlight style={styles.bottomButton} onPress={this._onToggleAddressModal} underlayColor={styleRaw.underlay.color}>
              <View style={styles.bottomButtonTextWrap}>
                <FAIcon style={styles.addressBookIcon} name="address-book-o" size={scale(18)} />
                <T style={styles.bottomButtonText}>{ADDRESS_TEXT}</T>
              </View>
            </TouchableHighlight>

            <TouchableHighlight style={styles.bottomButton} onPress={this._onToggleTorch} underlayColor={styleRaw.underlay.color}>
              <View style={styles.bottomButtonTextWrap}>
                <Ionicon style={styles.flashIcon} name="ios-flash" size={scale(24)} />
                <T style={styles.bottomButtonText}>{FLASH_TEXT}</T>
              </View>
            </TouchableHighlight>
          </View>
        </SceneWrapper>
        <SecondaryModal />
      </Fragment>
    )
  }

  _onToggleTorch = () => {
    this.props.toggleEnableTorch()
  }

  _onToggleAddressModal = () => {
    this.props.toggleAddressModal()
  }

  openSettingsTapped = () => {
    OpenAppSettings.open()
  }

  onBarCodeRead = (result: { data: string }) => {
    return this.props.qrCodeScanned(result.data)
  }

  renderCameraArea = () => {
    if (!this.props.scanEnabled) {
      return <View style={styles.cameraArea} />
    }

    if (this.props.cameraPermission === 'denied') {
      return (
        <View style={styles.cameraArea}>
          <Text style={styles.cameraPermissionDeniedText}>{DENIED_PERMISSION_TEXT}</Text>
          <TouchableHighlight style={styles.settingsButton} onPress={this.openSettingsTapped}>
            <Text style={styles.settingsButtonText}>{OPEN_SETTINGS_TEXT}</Text>
          </TouchableHighlight>
        </View>
      )
    }

    if (this.props.cameraPermission === 'authorized') {
      const flashMode = this.props.torchEnabled ? RNCamera.Constants.FlashMode.torch : RNCamera.Constants.FlashMode.off

      return (
        <RNCamera style={styles.cameraArea} captureAudio={false} flashMode={flashMode} onBarCodeRead={this.onBarCodeRead} type={RNCamera.Constants.Type.back}>
          <View style={styles.overlayTop}>
            <T style={styles.overlayTopText}>{HEADER_TEXT}</T>
          </View>
        </RNCamera>
      )
    }

    return (
      <View style={styles.cameraArea}>
        <ActivityIndicator size="large" color="#FFFFFF"/>
      </View>
    )
  }
}

export default Scan
