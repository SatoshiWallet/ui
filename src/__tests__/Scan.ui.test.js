// @flow
/* globals jest describe it expect */

import React from 'react'
import ShallowRenderer from 'react-test-renderer/shallow'

import { Scan } from '../components/scenes/ScanScene.js'

describe('Scan component', () => {
  it('should render with DENIED props', () => {
    const renderer = new ShallowRenderer()
    const props = {
      cameraPermission: 'denied',
      torchEnabled: false,
      scanEnabled: false,
      showToWalletModal: false,
      deepLinkPending: false,
      deepLinkUri: null,
      wallets: {},
      qrCodeScanned: jest.fn(),
      parseScannedUri: jest.fn(),
      markAddressDeepLinkDone: jest.fn(),
      toggleEnableTorch: jest.fn(),
      toggleAddressModal: jest.fn(),
      toggleScanToWalletListModal: jest.fn(),
      onSelectWallet: jest.fn()
    }
    const actual = renderer.render(<Scan {...props} />)

    expect(actual).toMatchSnapshot()
  })

  it('should render with AUTHORIZED props', () => {
    const renderer = new ShallowRenderer()
    const props = {
      cameraPermission: 'authorized',
      torchEnabled: false,
      scanEnabled: false,
      showToWalletModal: false,
      deepLinkPending: false,
      deepLinkUri: null,
      qrCodeScanned: jest.fn(),
      parseScannedUri: jest.fn(),
      markAddressDeepLinkDone: jest.fn(),
      toggleEnableTorch: jest.fn(),
      toggleAddressModal: jest.fn(),
      toggleScanToWalletListModal: jest.fn(),
      onSelectWallet: jest.fn(),
      wallets: {}
    }
    const actual = renderer.render(<Scan {...props} />)

    expect(actual).toMatchSnapshot()
  })
})
