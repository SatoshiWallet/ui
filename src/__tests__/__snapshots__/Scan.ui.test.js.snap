// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scan component should render with AUTHORIZED props 1`] = `
<React.Fragment>
  <SceneWrapper
    background="header"
    hasTabs={false}
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
        }
      }
    />
    <View
      style={
        Object {
          "flexDirection": "row",
          "paddingBottom": 15.450352112676057,
          "paddingLeft": 11.236619718309859,
          "paddingRight": 11.236619718309859,
          "paddingTop": 15.450352112676057,
        }
      }
    >
      <TouchableHighlight
        activeOpacity={0.85}
        delayPressOut={100}
        onPress={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#FFFFFF10",
            "borderRadius": 4.213732394366197,
            "flex": 1,
            "height": 70.22887323943662,
            "justifyContent": "center",
            "marginLeft": 1.4045774647887324,
            "marginRight": 1.4045774647887324,
          }
        }
        underlayColor="#0E4B75"
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "column",
              "justifyContent": "center",
            }
          }
        >
          <Icon
            allowFontScaling={false}
            name="address-book-o"
            size={25.28239436619718}
            style={
              Object {
                "color": "#FFFFFF",
                "fontSize": 22.473239436619718,
                "height": 22.473239436619718,
                "transform": Array [
                  Object {
                    "scaleX": -1,
                  },
                ],
              }
            }
          />
          <FormattedText
            style={
              Object {
                "backgroundColor": "transparent",
                "color": "#FFFFFF",
                "fontSize": 19.664084507042254,
                "opacity": 1,
              }
            }
          >
            Address
          </FormattedText>
        </View>
      </TouchableHighlight>
      <TouchableHighlight
        activeOpacity={0.85}
        delayPressOut={100}
        onPress={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#FFFFFF10",
            "borderRadius": 4.213732394366197,
            "flex": 1,
            "height": 70.22887323943662,
            "justifyContent": "center",
            "marginLeft": 1.4045774647887324,
            "marginRight": 1.4045774647887324,
          }
        }
        underlayColor="#0E4B75"
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "column",
              "justifyContent": "center",
            }
          }
        >
          <Icon
            allowFontScaling={false}
            name="ios-flash"
            size={33.709859154929575}
            style={
              Object {
                "color": "#FFFFFF",
                "fontSize": 30.900704225352115,
                "height": 25.28239436619718,
              }
            }
          />
          <FormattedText
            style={
              Object {
                "backgroundColor": "transparent",
                "color": "#FFFFFF",
                "fontSize": 19.664084507042254,
                "opacity": 1,
              }
            }
          >
            Flash
          </FormattedText>
        </View>
      </TouchableHighlight>
    </View>
  </SceneWrapper>
  <Connect(SecondaryModal) />
</React.Fragment>
`;

exports[`Scan component should render with DENIED props 1`] = `
<React.Fragment>
  <SceneWrapper
    background="header"
    hasTabs={false}
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
        }
      }
    />
    <View
      style={
        Object {
          "flexDirection": "row",
          "paddingBottom": 15.450352112676057,
          "paddingLeft": 11.236619718309859,
          "paddingRight": 11.236619718309859,
          "paddingTop": 15.450352112676057,
        }
      }
    >
      <TouchableHighlight
        activeOpacity={0.85}
        delayPressOut={100}
        onPress={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#FFFFFF10",
            "borderRadius": 4.213732394366197,
            "flex": 1,
            "height": 70.22887323943662,
            "justifyContent": "center",
            "marginLeft": 1.4045774647887324,
            "marginRight": 1.4045774647887324,
          }
        }
        underlayColor="#0E4B75"
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "column",
              "justifyContent": "center",
            }
          }
        >
          <Icon
            allowFontScaling={false}
            name="address-book-o"
            size={25.28239436619718}
            style={
              Object {
                "color": "#FFFFFF",
                "fontSize": 22.473239436619718,
                "height": 22.473239436619718,
                "transform": Array [
                  Object {
                    "scaleX": -1,
                  },
                ],
              }
            }
          />
          <FormattedText
            style={
              Object {
                "backgroundColor": "transparent",
                "color": "#FFFFFF",
                "fontSize": 19.664084507042254,
                "opacity": 1,
              }
            }
          >
            Address
          </FormattedText>
        </View>
      </TouchableHighlight>
      <TouchableHighlight
        activeOpacity={0.85}
        delayPressOut={100}
        onPress={[Function]}
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#FFFFFF10",
            "borderRadius": 4.213732394366197,
            "flex": 1,
            "height": 70.22887323943662,
            "justifyContent": "center",
            "marginLeft": 1.4045774647887324,
            "marginRight": 1.4045774647887324,
          }
        }
        underlayColor="#0E4B75"
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "column",
              "justifyContent": "center",
            }
          }
        >
          <Icon
            allowFontScaling={false}
            name="ios-flash"
            size={33.709859154929575}
            style={
              Object {
                "color": "#FFFFFF",
                "fontSize": 30.900704225352115,
                "height": 25.28239436619718,
              }
            }
          />
          <FormattedText
            style={
              Object {
                "backgroundColor": "transparent",
                "color": "#FFFFFF",
                "fontSize": 19.664084507042254,
                "opacity": 1,
              }
            }
          >
            Flash
          </FormattedText>
        </View>
      </TouchableHighlight>
    </View>
  </SceneWrapper>
  <Connect(SecondaryModal) />
</React.Fragment>
`;
